codegen/core/src/es/event/versioned.rs

```diff
@@ -0,0 +1,296 @@
+        let ty = &self.ident;
+        let (impl_gens, ty_gens, where_clause) = self.generics.split_for_impl();
+
+        let (event_name, event_ver) = (&self.event_name, &self.event_version);
+
+        quote! {
+            #[automatically_derived]
+            #[doc(hidden)]
+            impl #impl_gens ::arcana::codegen::UniqueEvents for #ty#ty_gens
+                 #where_clause
+            {
+                #[doc(hidden)]
+                const COUNT: usize = 1;
+            }
+
+            #[automatically_derived]
+            #[doc(hidden)]
+            impl #impl_gens #ty#ty_gens #where_clause {
+                #[doc(hidden)]
+                #[inline]
+                pub const fn __arcana_events() -> [(&'static str, u16); 1] {
+                    [(#event_name, #event_ver)]
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod spec {
+    use quote::quote;
+    use syn::parse_quote;
+
+    #[test]
+    fn derives_struct_impl() {
+        let input = parse_quote! {
+            #[event(name = "event", version = 1)]
+            struct Event;
+        };
+
+        let output = quote! {
+            #[automatically_derived]
+            impl ::arcana::es::event::Versioned for Event {
+                fn name() -> ::arcana::es::event::Name {
+                    "event"
+                }
+
+                fn version() -> ::arcana::es::event::Version {
+                    // SAFETY: Safe, as checked by proc macro in compile time.
+                    unsafe { ::arcana::es::event::Version::new_unchecked(1) }
+                }
+            }
+
+            #[automatically_derived]
+            #[doc(hidden)]
+            impl ::arcana::codegen::UniqueEvents for Event {
+                #[doc(hidden)]
+                const COUNT: usize = 1;
+            }
+
+            #[automatically_derived]
+            #[doc(hidden)]
+            impl Event {
+                #[doc(hidden)]
+                pub const fn __arcana_events() -> [(&'static str, u16); 1] {
+                    [("event", 1)]
+                }
+            }
+        };
+
+        assert_eq!(
+            super::derive(input).unwrap().to_string(),
+            output.to_string(),
+        );
+    }
+
+    #[test]
+    fn name_arg_is_required() {
+        let input = parse_quote! {
+            #[event(ver = 1)]
+            struct Event;
+        };
+
+        let err = super::derive(input).unwrap_err();
+
+        assert_eq!(
+            format!("{}", err),
+            "`name` argument of `#[event]` attribute is expected to be \
+             present, but is absent",
+        );
+    }
+
+    #[test]
+    fn version_arg_is_required() {
+        let input = parse_quote! {
+            #[event(name = "event")]
+            struct Event;
+        };
+
+        let err = super::derive(input).unwrap_err();
+
+        assert_eq!(
+            format!("{}", err),
+            "either `version` or `ver` argument of `#[event]` attribute is \
```

Comment:
@tyranron it looks like error message isn't deterministic. Sometimes I get `either 'ver' or 'version' argument of '#[event]' attribute is expected to be present, but is absent`, so `ver` and `version` are switched (even on 2 CI runs different jobs failed). As I understand this happens because `synthez` uses `HashSet<syn::Ident>` under the hood. `HashSet` docs are saying that there is requirement

> k1 == k2 -> hash(k1) == hash(k2)

And for `syn::Ident` that promise isn't held up as I understand. [`PartialEq` impl](https://docs.rs/proc-macro2/1.0.28/src/proc_macro2/lib.rs.html#949-953) compares ident itself and it's `Span`, while [`Hash` impl](https://docs.rs/proc-macro2/1.0.28/src/proc_macro2/lib.rs.html#978-982) simply hashes ident, ignoring it's `Span`. This can be easily fixed specifically for error message generation, but I'm not sure if this thing can cause bigger ploblems.